### PR TITLE
Adding cache key generator + test coverage

### PR DESCRIPTION
**Pull request by burzum619**
Adding cache key generator to TDS_Common. This key gen will generate keys based on class name, method name, and parameter hashcodes. Test coverage is included for the key generator.

### CODE REVIEW

File: common/pom.xml [hunk 1]
```diff
@@ -97,6 +97,11 @@
             <artifactId>spring-context-support</artifactId>
             <version>4.3.2.RELEASE</version>
         </dependency>
+        <dependency>
+            <groupId>org.springframework</groupId>
+            <artifactId>spring-test</artifactId>
+            <version>4.3.2.RELEASE</version>
```
Comment (millerts) [inline common/pom.xml:103]: I think this should have a `<scope>test</scope>` since it is only needed for testing.

File: common/src/main/java/tds/common/cache/CacheKeyGenerator.java [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+ * A custom key generator for caching
+ */
+public class CacheKeyGenerator implements KeyGenerator {
+    @Override
+    public Object generate(Object target, Method method, Object... params) {
+        return this.generate(target.getClass().getName(), method.getName(), params);
+    }
+
+    public String generate(String className, String methodName, Object... params) {
+        StringBuilder sb = new StringBuilder();
```
Comment (millerts) [inline common/src/main/java/tds/common/cache/CacheKeyGenerator.java:17]: You can chain these so it'd be `StringBuilder sb = new StringBuilder(className).append(".").append(methodName);`

File: common/src/main/java/tds/common/cache/CacheKeyGenerator.java [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+        sb.append(className);
+        sb.append(".");
+        sb.append(methodName);
+
+        for (int i = 0; i < params.length; ++i) {
```
Comment (millerts) [inline common/src/main/java/tds/common/cache/CacheKeyGenerator.java:22]: This can be changed to be `for(Object obj : params)`.  `i` isn't necessary.

File: common/src/main/java/tds/common/cache/CacheKeyGenerator.java [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+            Object obj = params[i];
+
+            if(obj != null) {
+                sb.append(obj.hashCode());
+            }
+
+            sb.append(":");
```
Comment (millerts) [inline common/src/main/java/tds/common/cache/CacheKeyGenerator.java:29]: Just for my clarification:

`SomethingService.findSomethingByKeyAndName("name", null);`

will result in

`"SomethingService.findSomethingByKeyAndNamename::"`

Comment (millerts) [reply common/src/main/java/tds/common/cache/CacheKeyGenerator.java:29]: I know it doesn't matter for uniqueness but should we add a `:` or some other character between the method name and the first parameter?

File: common/src/main/java/tds/common/configuration/CacheConfiguration.java [hunk 5]
```diff
@@ -37,23 +41,25 @@ public CacheManager cacheManager() {
     }
 
     @Bean
-    @Qualifier(CacheType.SHORT_TERM)
+    public KeyGenerator keyGenerator() {
+        return new CacheKeyGenerator();
+    }
+
+    @Bean
```
Comment (millerts) [inline common/src/main/java/tds/common/configuration/CacheConfiguration.java:48]: So you didn't need to name the bean [CacheType.SHORT_TERM](https://github.com/SmarterApp/TDS_Common/blob/develop/common/src/main/java/tds/common/configuration/CacheType.java#L9)?

**Review by millerts — CHANGES_REQUESTED** [review]
Some minor changes

File: common/src/main/java/tds/common/configuration/CacheConfiguration.java (hunk 5 above)
Comment (burzum619) [reply common/src/main/java/tds/common/configuration/CacheConfiguration.java:48]: Nope. Behind the scenes Spring will look at the cachename of the cache, not the bean id.

File: common/src/main/java/tds/common/cache/CacheKeyGenerator.java (hunk 2 above)
Comment (burzum619) [reply common/src/main/java/tds/common/cache/CacheKeyGenerator.java:17]: I blame @jtreuting

File: common/src/main/java/tds/common/cache/CacheKeyGenerator.java (hunk 4 above)
Comment (burzum619) [reply common/src/main/java/tds/common/cache/CacheKeyGenerator.java:29]: Sounds good to me. I'll add some test coverage for null/empty strings.

File: common/src/main/java/tds/common/cache/CacheKeyGenerator.java (hunk 2 above)
Comment (jtreuting) [reply common/src/main/java/tds/common/cache/CacheKeyGenerator.java:17]: Definitely, @burzum619 just pulled this over as is.

File: common/src/main/java/tds/common/cache/CacheKeyGenerator.java (hunk 4 above)
Comment (jtreuting) [reply common/src/main/java/tds/common/cache/CacheKeyGenerator.java:29]: Yeah good catch.  Looks like null and empty string will result in same cache key when they shouldn't.  Let's add a specific test case for that

File: common/src/main/java/tds/common/cache/CacheKeyGenerator.java (hunk 2 above)
Comment (burzum619) [reply common/src/main/java/tds/common/cache/CacheKeyGenerator.java:17]: Just kidding :) Shoulda cleaned up a bit instead of copy-pasta-ing

File: common/src/main/java/tds/common/cache/CacheKeyGenerator.java (hunk 4 above)
Comment (burzum619) [reply common/src/main/java/tds/common/cache/CacheKeyGenerator.java:29]: Done.

`SomethingService.findSomethingByKeyAndName("name", null);`

Now results in 
`"SomethingService.findSomethingByKeyAndName:name::"`

and `SomethingService.findSomethingByKeyAndName("name", "");`

results in
`"SomethingService.findSomethingByKeyAndName:name:0:"`
